pandas/tests/reshape/test_from_dummies.py

```diff
@@ -0,0 +1,441 @@
+    dummies = DataFrame(
+        {
+            "C": [1, 2, 3],
+            "col1_a": [1, 0, 1],
+            "col1_b": [0, 1, 0],
+            "col2_a": [0, 1, 0],
+            "col2_b": [1, 0, 0],
+            "col2_c": [0, 0, 1],
+        },
+    )
+    expected = DataFrame({"C": [1, 2, 3], "A": ["b", "a", "c"], "B": ["a", "b", "a"]})
+    result = from_dummies(dummies, variables={"col2": "A", "col1": "B"})
+    assert all(result == expected)
+
+
+def test_from_dummies_to_df_variable_dict_not_complete():
+    dummies = DataFrame(
+        {
+            "C": [1, 2, 3],
+            "col1_a": [1, 0, 1],
+            "col1_b": [0, 1, 0],
+            "col2_a": [0, 1, 0],
+            "col2_b": [1, 0, 0],
+            "col2_c": [0, 0, 1],
+        },
+    )
+    with pytest.raises(
+        ValueError,
+        match=(
+            r"Length of 'variables' \(1\) did not match "
+            r"the length of the columns being encoded \(2\)."
+        ),
+    ):
+        from_dummies(dummies, variables={"col1": "A"})
+
+
+def test_from_dummies_to_df_prefix_sep_list():
+    dummies = DataFrame(
+        {
+            "C": [1, 2, 3],
+            "col1_a": [1, 0, 1],
+            "col1_b": [0, 1, 0],
+            "col2-a": [0, 1, 0],
+            "col2-b": [1, 0, 0],
+            "col2-c": [0, 0, 1],
+        },
+    )
+    expected = DataFrame(
+        {"C": [1, 2, 3], "col1": ["a", "b", "a"], "col2": ["b", "a", "c"]}
+    )
+    result = from_dummies(dummies, prefix_sep=["_", "-"])
+    assert all(result == expected)
+
+
+def test_from_dummies_to_df_prefix_sep_dict():
+    dummies = DataFrame(
+        {
+            "C": [1, 2, 3],
+            "col1_a-a": [1, 0, 1],
+            "col1_b-a": [0, 1, 0],
+            "col2-a_a": [0, 1, 0],
+            "col2-b_b": [1, 0, 0],
+            "col2-c_c": [0, 0, 1],
+        },
+    )
+    expected = DataFrame(
+        {"C": [1, 2, 3], "col1": ["a-a", "b-b", "a-a"], "col2": ["b_b", "a_a", "c_c"]}
+    )
+    result = from_dummies(
+        dummies,
+        prefix_sep={
+            "col1": "_",
+            "col2": "-",
+        },
+    )
+    assert all(result == expected)
+
+
+def test_from_dummies_to_df_dummy_na():
+    dummies = DataFrame(
+        {
+            "C": [1, 2, 3],
+            "col1_a": [1, 0, 0],
+            "col1_b": [0, 1, 0],
+            "col1_NaN": [0, 0, 1],
+            "col2_a": [0, 1, 0],
+            "col2_b": [0, 0, 0],
+            "col2_c": [0, 0, 1],
+            "col2_NAN": [1, 0, 0],
+        },
+    )
+    expected = DataFrame(
+        {"C": [1, 2, 3], "col1": ["a", "b", np.nan], "col2": [np.nan, "a", "c"]}
+    )
+    result = from_dummies(dummies, dummy_na=True)
+    assert all(result["C"] == expected["C"])
+    assert all(result["col1"][:2] == expected["col1"][:2])
+    assert all(result["col2"][1:] == expected["col2"][1:])
+    assert result["col1"][2] is expected["col1"][2]
+    assert result["col2"][1] is expected["col2"][1]
```

Comment:
There must be a more elegant way of testing this equality but the different dtypes seem to make it tricky if I am not missing something.